File: DigitalImageProcessing/python/midden_salt.py
#!/usr/bin/env python3
# encoding: utf-8


"""
@Funciton: 椒盐噪声; 中值滤波(平滑)-非线性滤波器
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-09-24
"""

import sys
import random
import cv2 as cv
import numpy as np


def salt(image, number):
    rows, cols = image.shape
    saltImg = np.copy(image)
    for idx in range(number):
        rand_row = random.randint(0, rows-1)
        rand_col = random.randint(0, cols-1)
        saltImg[rand_row][rand_col] = 255

    return saltImg


def medianBlur(image, winSize):
    rows, cols = image.shape
    winH, winW = winSize
    halfWinH = (winH - 1) // 2
    halfWinW = (winW - 1) // 2
    medianBlurImg = np.zeros(image.shape, image.dtype)
    for idx_row in range(rows):
        for idx_col in range(cols):
            # 边界判断
            rTop = 0 if idx_row - halfWinH < 0 else idx_row - halfWinH
            rBottom = rows - 1 if idx_row + halfWinH > rows - 1 else idx_row + halfWinH
            cLeft = 0 if idx_col - halfWinW < 0 else idx_col - halfWinW
            cRight = cols - 1 if idx_col + halfWinW > cols - 1 else idx_col + halfWinW
            # 取领域
            region = image[rTop:rBottom+1, cLeft:cRight+1]
            # 求中值
            medianBlurImg[idx_row][idx_col] = np.median(region)

    return medianBlurImg


# --------------------------
if __name__ == "__main__":
    if len(sys.argv) > 1:
        image = cv.imread(sys.argv[1], 0)
        if image is None:
            print(f"Error: no such file or dictory.")
        cv.imshow("OriginImage", image)

        # 0. 添加椒盐噪声
        saltImg = salt(image, 142)
        cv.imshow("saltImg", saltImg)

        # 1. 中值滤波
        medianImg = medianBlur(saltImg, (5, 5))
        cv.imshow("medianImg", medianImg.astype(np.uint8))

        cv.waitKey(0)
        cv.destroyAllWindows()

    else:
        print(f"Usage: python histogram imageFile.")

File: DigitalImageProcessing/cpp/midden_salt.cpp
/**
 * @File    : midden_salt.cpp
 * @Brief   : 椒盐噪声; 中值滤波(平滑)-非线性滤波器
 * @Author  : Wei Li
 * @Date    : 2021-09-24
*/

#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

cv::Mat medianSmooth(cv::Mat img, cv::Size winSize, int borderType=cv::BORDER_DEFAULT)
{
    CV_Assert(img.type() == CV_8UC1);
    int H = winSize.height;
    int W = winSize.width;
    CV_Assert(H > 0 && W > 0);
    CV_Assert(H % 2 == 1 && W % 2 == 1);
    int h = (H - 1) / 2;
    int w = (W - 1) / 2;

    cv::Mat paddImg;
    cv::copyMakeBorder(img, paddImg, h, h, w, w, borderType);

    int rows = img.rows;
    int cols = img.cols;
    cv::Mat medianImg(img.size(), CV_8UC1);

    int i = 0, j = 0;
    // 中数的位置
    int index = (H*W - 1) / 2;
    for (size_t r = 0; r < h + rows; ++r)
    {
        for (size_t c = 0; c < w + cols; ++c)
        {
            // 取以当前位置为中心，大小为 winSize 的邻域
            cv::Mat region = paddImg(cv::Rect(c-w, r-h, W, H)).clone();
            // 将该邻域转换为行矩阵
            region = region.reshape(1, 1);
            // 排序
            cv::sort(region, region, cv::SORT_EVERY_ROW);
            // 取中数
            uchar medianValue = region.at<uchar>(0, index);
            medianImg.at<uchar>(i, j) = medianValue;
            ++j;
        }
        ++i;
        j = 0;
    }
    return medianImg;
}

// -------------------------------
int main(int argc, char **argv)
{
    if (argc > 1)
    {
        cv::Mat image = cv::imread(argv[1], 0);
        if (!image.data)
        {
            std::cout << "Error: reading image unsuccesfully." << std::endl;
            return -1;
        }
        cv::imshow("OriginImage", image);

        // 1. OpenCV 函数进行中值滤波器
        cv::Mat median_blur_img = medianSmooth(image, cv::Size(5, 5));
        cv::imshow("median_blur_img", median_blur_img);

        // 2. 利用 OpenCV 提供的函数，计算效率
        cv::Mat median_dst;
        cv::medianBlur(image, median_dst, 5);
        cv::imshow("medianImg", median_dst);

        cv::waitKey(0);
        cv::destroyAllWindows();
        return 0;
    }
    else
    {
        std::cout << "Usage: OpenCV gauss convolution image." << std::endl;
        return -1;
    }
}
